# 2015 NYC Street Tree Census dataset
Explore, clean and prepare dataset for a machine learning model

Current main objective: to understand the determinant factors of trees health in New York City.

## 1. Import modules

In [1]:
import numpy as np
import pandas as pd

## 2. Import dataset

In [2]:
df = pd.read_csv('assets/data_100000.csv')

## 3. Explore dataset

In [3]:
print("0. The five first rows:")
df.head()

0. The five first rows:


created_at  tree_id  block_id  \
0  08/27/2015   180683    348711   
1  09/03/2015   200540    315986   
2  09/05/2015   204026    218365   
3  09/05/2015   204337    217969   
4  08/30/2015   189565    223043   

                                        the_geom  tree_dbh  stump_diam  \
0  POINT (-73.84421521958048 40.723091773924274)         3           0   
1   POINT (-73.81867945834878 40.79411066708779)        21           0   
2  POINT (-73.93660770459083 40.717580740099116)         3           0   
3  POINT (-73.93445615919741 40.713537494833226)        10           0   
4   POINT (-73.97597938483258 40.66677775537875)        21           0   

  curb_loc status health                           spc_latin  ... st_assem  \
0   OnCurb  Alive   Fair                         Acer rubrum  ...       28   
1   OnCurb  Alive   Fair                   Quercus palustris  ...       27   
2   OnCurb  Alive   Good  Gleditsia triacanthos var. inermis  ...       50   
3   OnCurb  Alive   Good  Gleditsia triacanthos var. inermis  ...       53   
4   OnCurb  Alive   Good                     Tilia americana  ...       44   

  st_senate   nta            nta_name  boro_ct     state   latitude  \
0        16  QN17        Forest Hills  4073900  New York  40.723092   
1        11  QN49          Whitestone  4097300  New York  40.794111   
2        18  BK90   East Williamsburg  3044900  New York  40.717581   
3        18  BK90   East Williamsburg  3044900  New York  40.713537   
4        21  BK37  Park Slope-Gowanus  3016500  New York  40.666778   

   longitude          x_sp           y_sp  
0 -73.844215  1.027431e+06  202756.768749  
1 -73.818679  1.034456e+06  228644.837379  
2 -73.936608  1.001823e+06  200716.891267  
3 -73.934456  1.002420e+06  199244.253136  
4 -73.975979  9.909138e+05  182202.425999  

[5 rows x 42 columns]

In [4]:
print("1. General info of dataset:")
df.info()

1. General info of dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 42 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   created_at  100000 non-null  object 
 1   tree_id     100000 non-null  int64  
 2   block_id    100000 non-null  int64  
 3   the_geom    100000 non-null  object 
 4   tree_dbh    100000 non-null  int64  
 5   stump_diam  100000 non-null  int64  
 6   curb_loc    100000 non-null  object 
 7   status      100000 non-null  object 
 8   health      95007 non-null   object 
 9   spc_latin   95008 non-null   object 
 10  spc_common  95008 non-null   object 
 11  steward     95008 non-null   object 
 12  guards      95008 non-null   object 
 13  sidewalk    95008 non-null   object 
 14  user_type   100000 non-null  object 
 15  problems    95008 non-null   object 
 16  root_stone  100000 non-null  object 
 17  root_grate  100000 non-null  object 
 18  root_other  10000

In [5]:
print(f"2. Number of rows: {len(df)}")
print(f"3. Number of original features: {len(df.columns)}")

2. Number of rows: 100000
3. Number of original features: 42


### Exploring missing values

In [6]:
print("4. Number of missing values in each original feature:")
df.isnull().sum()

4. Number of missing values in each original feature:


created_at       0
tree_id          0
block_id         0
the_geom         0
tree_dbh         0
stump_diam       0
curb_loc         0
status           0
health        4993
spc_latin     4992
spc_common    4992
steward       4992
guards        4992
sidewalk      4992
user_type        0
problems      4992
root_stone       0
root_grate       0
root_other       0
trnk_wire        0
trnk_light       0
trnk_other       0
brnch_ligh       0
brnch_shoe       0
brnch_othe       0
address          0
zipcode          0
zip_city         0
cb_num           0
borocode         0
boroname         0
cncldist         0
st_assem         0
st_senate        0
nta              0
nta_name         0
boro_ct          0
state            0
latitude         0
longitude        0
x_sp             0
y_sp             0
dtype: int64

In [7]:
print("5. Data type of each original feature:")
df.dtypes

5. Data type of each original feature:


created_at     object
tree_id         int64
block_id        int64
the_geom       object
tree_dbh        int64
stump_diam      int64
curb_loc       object
status         object
health         object
spc_latin      object
spc_common     object
steward        object
guards         object
sidewalk       object
user_type      object
problems       object
root_stone     object
root_grate     object
root_other     object
trnk_wire      object
trnk_light     object
trnk_other     object
brnch_ligh     object
brnch_shoe     object
brnch_othe     object
address        object
zipcode         int64
zip_city       object
cb_num          int64
borocode        int64
boroname       object
cncldist        int64
st_assem        int64
st_senate       int64
nta            object
nta_name       object
boro_ct         int64
state          object
latitude      float64
longitude     float64
x_sp          float64
y_sp          float64
dtype: object

In [8]:
print("6. Number of unique values of each original feature:")
print("NaN non included")
df.nunique()

6. Number of unique values of each original feature:
NaN non included


created_at       146
tree_id       100000
block_id       19472
the_geom       99962
tree_dbh          77
stump_diam        75
curb_loc           2
status             3
health             3
spc_latin        125
spc_common       125
steward            4
guards             4
sidewalk           2
user_type          3
problems         151
root_stone         2
root_grate         2
root_other         2
trnk_wire          2
trnk_light         2
trnk_other         2
brnch_ligh         2
brnch_shoe         2
brnch_othe         2
address        66793
zipcode          185
zip_city          47
cb_num            59
borocode           5
boroname           5
cncldist          51
st_assem          65
st_senate         26
nta              188
nta_name         188
boro_ct         1652
state              1
latitude       99799
longitude      99771
x_sp           99929
y_sp           99940
dtype: int64

In [10]:
print("7. Unique values of original features with less than 204 unique values:")
for column in df.columns:
    if df[column].nunique() < 200:
        print(f"#################### {column} ####################")
        print(df[column].unique())
        print()

8  54  74  64  49  57  77 118  51 109  67  80 134  78
  79  69  76]

#################### curb_loc ####################
['OnCurb' 'OffsetFromCurb']

#################### status ####################
['Alive' 'Dead' 'Stump']

#################### health ####################
['Fair' 'Good' 'Poor' nan]

#################### spc_latin ####################
['Acer rubrum' 'Quercus palustris' 'Gleditsia triacanthos var. inermis'
 'Tilia americana' 'Platanus x acerifolia' 'Ginkgo biloba'
 'Quercus phellos' 'Acer pseudoplatanus' 'Acer ginnala' 'Acer campestre'
 'Ulmus americana' 'Fraxinus' 'Malus' 'Acer saccharinum' 'Corylus colurna'
 'Prunus serotina' 'Juniperus virginiana' 'Acer platanoides'
 'Liriodendron tulipifera' 'Quercus acutissima' 'Styphnolobium japonicum'
 'Quercus bicolor' 'Chionanthus retusus' 'Magnolia grandiflora'
 'Liquidambar styraciflua' 'Pyrus calleryana' 'Quercus coccinea'
 'Chamaecyparis thyoides' 'Quercus velutina' 'Zelkova serrata'
 'Quercus alba' 'Aesculus glabra' 'Quercu

In [11]:
print("8. Statistical info of each original feature:")
df.describe(include='all').T

8. Statistical info of each original feature:


count unique                                           top  \
created_at    100000    146                                    09/23/2015   
tree_id     100000.0    NaN                                           NaN   
block_id    100000.0    NaN                                           NaN   
the_geom      100000  99962  POINT (-73.95680061508556 40.61596593024687)   
tree_dbh    100000.0    NaN                                           NaN   
stump_diam  100000.0    NaN                                           NaN   
curb_loc      100000      2                                        OnCurb   
status        100000      3                                         Alive   
health         95007      3                                          Good   
spc_latin      95008    125                         Platanus x acerifolia   
spc_common     95008    125                              London planetree   
steward        95008      4                                          None   
guards         95008      4                                          None   
sidewalk       95008      2                                      NoDamage   
user_type     100000      3                                     Volunteer   
problems       95008    151                                          None   
root_stone    100000      2                                            No   
root_grate    100000      2                                            No   
root_other    100000      2                                            No   
trnk_wire     100000      2                                            No   
trnk_light    100000      2                                            No   
trnk_other    100000      2                                            No   
brnch_ligh    100000      2                                            No   
brnch_shoe    100000      2                                            No   
brnch_othe    100000      2                                            No   
address       100000  66793                           28-055 ULMER STREET   
zipcode     100000.0    NaN                                           NaN   
zip_city      100000     47                                      Brooklyn   
cb_num      100000.0    NaN                                           NaN   
borocode    100000.0    NaN                                           NaN   
boroname      100000      5                                        Queens   
cncldist    100000.0    NaN                                           NaN   
st_assem    100000.0    NaN                                           NaN   
st_senate   100000.0    NaN                                           NaN   
nta           100000    188                                          QN43   
nta_name      100000    188                                     Bellerose   
boro_ct     100000.0    NaN                                           NaN   
state         100000      1                                      New York   
latitude    100000.0    NaN                                           NaN   
longitude   100000.0    NaN                                           NaN   
x_sp        100000.0    NaN                                           NaN   
y_sp        100000.0    NaN                                           NaN   

              freq            mean             std            min  \
created_at    3916             NaN             NaN            NaN   
tree_id        NaN    233739.45083    42482.723266            7.0   
block_id       NaN    289398.95714   122324.350099       100002.0   
the_geom        17             NaN             NaN            NaN   
tree_dbh       NaN        11.26171        8.544646            0.0   
stump_diam     NaN         0.47271        3.457305            0.0   
curb_loc     96129             NaN             NaN            NaN   
status       95008             NaN             NaN            NaN   
health       72340             NaN             NaN            NaN   
spc_latin    

In [12]:
print("9. Duplicates in the dataset? ")
df.duplicated().any()

9. Duplicates in the dataset? 


False

### Explore deeper missing values

health        4993
spc_latin     4992
spc_common    4992
steward       4992
guards        4992
sidewalk      4992
user_type        0
problems      4992

In [34]:
print("Original features with missing values:")
cols_with_nan = df.loc[:, df.isnull().sum() > 0].columns
cols_with_nan

Original features with missing values:


Index(['health', 'spc_latin', 'spc_common', 'steward', 'guards', 'sidewalk',
       'problems'],
      dtype='object')

In [51]:
# Add status to the list of features with missing values
cols_with_nan_and_status = cols_with_nan.values.tolist()
cols_with_nan_and_status.append("status")
cols_with_nan_and_status

['health',
 'spc_latin',
 'spc_common',
 'steward',
 'guards',
 'sidewalk',
 'problems',
 'status']

In [63]:
# Count number of values in each feature with nan per status 
df[cols_with_nan_and_status].groupby(["status"]).count()

health  spc_latin  spc_common  steward  guards  sidewalk  problems
status                                                                    
Alive    95007      95008       95008    95008   95008     95008     95008
Dead         0          0           0        0       0         0         0
Stump        0          0           0        0       0         0         0

In [61]:
# See example where `health` is null and status different than "Dead" or "Stump"
df.loc[(df["health"].isnull()) & (~df["status"].isin(["Dead", "Stump"]))].T

32794
created_at                                    09/21/2015
tree_id                                           245041
block_id                                          413012
the_geom    POINT (-74.21641202587655 40.54859739520807)
tree_dbh                                              16
stump_diam                                             0
curb_loc                                          OnCurb
status                                             Alive
health                                               NaN
spc_latin                         Fraxinus pennsylvanica
spc_common                                     green ash
steward                                             None
guards                                              None
sidewalk                                          Damage
user_type                               TreesCount Staff
problems                                            None
root_stone                                            No
root_grate                                            No
root_other                                            No
trnk_wire                                             No
trnk_light                                            No
trnk_other                                            No
brnch_ligh                                            No
brnch_shoe                                            No
brnch_othe                                            No
address                                84 LUCILLE AVENUE
zipcode                                            10309
zip_city                                   Staten Island
cb_num                                               503
borocode                                               5
boroname                                   Staten Island
cncldist                                              51
st_assem                                              62
st_senate                                             24
nta                                                 SI32
nta_name                               Rossville-Woodrow
boro_ct                                          5020801
state                                           New York
latitude                                       40.548597
longitude                                     -74.216412
x_sp                                       924106.880773
y_sp                                       139219.632044

In [78]:
# Count how many trees per `status`
df["status"].value_counts()

Alive    95008
Stump     2831
Dead      2161
Name: status, dtype: int64

In [79]:
# Count number of values per `status` and `health`
df[["status", "health"]].value_counts()

status  health
Alive   Good      72340
        Fair      17266
        Poor       5401
dtype: int64

### Observation notes
#### Missing values
* When `status` = "Dead" or "Stump":
    - 'health', 'spc_latin', 'spc_common', 'steward', 'guards', 'sidewalk',
       'problems' are empty.
    - Potential issue: if we look at trees that are alive, a lot of them are considered healthy ("Good") and a few "Poor". Maybe due to when the trees are "Dead" or "Stump", no perception of health is given.
    
    => Question: for feature `health`, should we fill in "poor" when `status` is "Dead" or "Stump"? No, that could distort the information about `health` user's perception.

    => Decision: 
        - for feature `health`, we are going to fill in "Dead" or "Stump" when `status` is "Dead" or "Stump" as this is a factual and correct health description.
        - for the other features with missing values: we are going to populated them with the text "NA".


* 1 particular case:
    - There is one entry (=row) where `health` is null and `status` different than "Dead" or "Stump".

    => Decision: drop this particular case. This won't have a big impact statistically.
    

### Data types
* `created_at` is object (= string)

    => Decision: Convert it into a *datetime64* feature




## 4. Clean Dataset

### Fix missing values

In [85]:
# Fill in `health` with `status` when `status` is "Dead" or "Stump" and `health` is null
condition = (df["status"].isin(["Dead", "Stump"]) & df["health"].isnull())
df["health"][(condition)] = df["status"][(condition)]

In [88]:
# Check if `health` is correctly mapped with `status` "Dead" or "Stump"
df[["health", "status"]][df["status"].isin(["Dead", "Stump"])].head()

health status
630   Dead   Dead
631  Stump  Stump
634  Stump  Stump
635   Dead   Dead
636  Stump  Stump

In [100]:
# Original features with missing values
cols_with_nan = df.loc[:, df.isnull().sum() > 0].columns
for column in cols_with_nan:
    df[column][df[column].isnull()] = "NA"

In [102]:
# Check if original features with missing values have been correctly populated
df[cols_with_nan][df["status"].isin(["Dead", "Stump"])].head()

spc_latin spc_common steward guards sidewalk problems
630        NA         NA      NA     NA       NA       NA
631        NA         NA      NA     NA       NA       NA
634        NA         NA      NA     NA       NA       NA
635        NA         NA      NA     NA       NA       NA
636        NA         NA      NA     NA       NA       NA

In [101]:
# Check if there are any mising values
df.isnull().sum()

created_at    0
the_geom      0
tree_dbh      0
stump_diam    0
curb_loc      0
status        0
health        0
spc_latin     0
spc_common    0
steward       0
guards        0
sidewalk      0
user_type     0
problems      0
root_stone    0
root_grate    0
root_other    0
trnk_wire     0
trnk_light    0
trnk_other    0
brnch_ligh    0
brnch_shoe    0
brnch_othe    0
address       0
zipcode       0
zip_city      0
cb_num        0
borocode      0
boroname      0
cncldist      0
st_assem      0
st_senate     0
nta           0
nta_name      0
boro_ct       0
state         0
latitude      0
longitude     0
x_sp          0
y_sp          0
dtype: int64

### Fix data types

In [82]:
# Convert `created_at` into a datetime64 feature
df["created_at"] = pd.to_datetime(df["created_at"])

In [83]:
# Check if `created_at` is correctly converted into a datetime64 feature
df["created_at"].head()

0   2015-08-27
1   2015-09-03
2   2015-09-05
3   2015-09-05
4   2015-08-30
Name: created_at, dtype: datetime64[ns]

### Feature selection

In [84]:
# Drop the `tree_id` and `block_id` features as they won't be useful for predicting the outcome.
df.drop(["tree_id", "block_id"], axis=1, inplace=True)

In [97]:
# Drop the particular case where `health` is null and status different than "Dead" or "Stump"
alive_no_status_idx = df.loc[(df["health"].isnull()) & (~df["status"].isin(["Dead", "Stump"]))].index
alive_no_status_idx
df.drop(alive_no_status_idx, axis=0, inplace=True)

In [104]:
# Quick and dirty version
# we'll only select the most important features where there is no missing value.
no_nan_cols = df.loc[:, df.isnull().sum() == 0].columns
df_good_enough = df[no_nan_cols]

In [105]:
# Export the dataframe into data_clean_GOOD_ENOUGH file
df_good_enough.to_csv(r'assets/data_clean_GOOD_ENOUGH.csv', index = False)